sympy/matrices/eigen.py

```diff
@@ -1228,40 +1255,18 @@ def nullity_chain(val, algebraic_multiplicity):
 
         return ret
 
-    def blocks_from_nullity_chain(d):
-        """Return a list of the size of each Jordan block.
-        If d_n is the nullity of E**n, then the number
-        of Jordan blocks of size n is
-
-            2*d_n - d_(n-1) - d_(n+1)"""
-
-        # d[0] is always the number of columns, so skip past it
-        mid = [2*d[n] - d[n - 1] - d[n + 1] for n in range(1, len(d) - 1)]
-        # d is assumed to plateau with "d[ len(d) ] == d[-1]", so
-        # 2*d_n - d_(n-1) - d_(n+1) == d_n - d_(n-1)
-        end = [d[-1] - d[-2]] if len(d) > 1 else [d[0]]
-
-        return mid + end
-
-    def pick_vec(small_basis, big_basis):
-        """Picks a vector from big_basis that isn't in
-        the subspace spanned by small_basis"""
-
-        if len(small_basis) == 0:
-            return big_basis[0]
-
-        for v in big_basis:
-            _, pivots = M.hstack(*(small_basis + [v])).echelon_form(
-                    with_pivots=True)
-
-            if pivots[-1] == len(small_basis):
-                return v
-
     # roots doesn't like Floats, so replace them with Rationals
     if has_floats:
         from sympy.simplify import nsimplify
         mat = mat.applyfunc(lambda x: nsimplify(x, rational=True))
 
+    # check if matrix is rational
+    has_rationals = all(num.is_rational for num in mat)
+
+    if has_rationals:
+        jordan_result = jordan_form_rational_matrix(mat, calc_transform)
+        return restore_floats2(*jordan_result) if calc_transform else restore_floats1(jordan_result)
```

Comment:
I would use `.iter_values` to check for rational numbers:
```python
In [6]: M = randMatrix(5, percent=10)

In [7]: M
Out[7]: 
⎡0  0  0  6  0 ⎤
⎢              ⎥
⎢0  0  0  0  0 ⎥
⎢              ⎥
⎢0  0  0  0  0 ⎥
⎢              ⎥
⎢0  0  0  0  43⎥
⎢              ⎥
⎣0  0  0  0  0 ⎦

In [8]: M.iter_values()
Out[8]: <map at 0x7907cc128280>

In [9]: list(M.iter_values())
Out[9]: [6, 43]
```
The way that you have written this will not work well for static typing. I would write this as:
```python
if calc_transform:
    P, J = jordan_form_rational_matrix(mat, calc_transform=True)
    return restore_floats2(P, J)
else:
    J = jordan_form_rational_matrix(mat, calc_transform=False)
    return restore_floats1(J)
```